src/test/state_machine.zig

```diff
@@ -23,21 +26,29 @@ pub const StateMachine = struct {
         realtime: i64,
         operation: Operation,
         input: []u8,
-    ) void {
-        _ = state_machine;
-        _ = realtime;
+    ) u64 {
         _ = operation;
         _ = input;
+
         // TODO: use realtime in some way to test the system
+        state_machine.prepare_timestamp = std.math.max(
+            std.math.max(state_machine.prepare_timestamp, state_machine.commit_timestamp) + 1,
+            @intCast(u64, realtime));
+        return state_machine.prepare_timestamp;
     }
 
     pub fn commit(
         state_machine: *StateMachine,
         client: u128,
         operation: Operation,
+        timestamp: u64,
         input: []const u8,
         output: []u8,
     ) usize {
+        defer {
+            assert(state_machine.commit_timestamp <= timestamp);
+            state_machine.commit_timestamp = timestamp;
+        }
```

Comment:
Seeing how we have to duplicate the `prepare/commit_timestamp` logic for all state machines (TB's and our test state machine here), let's rather move the timestamp assignment and assertions outside of `StateMachine` to be taken care of by `Replica`.

The principle here is to keep the design as a library, to make the state machine really easy to stub out, without having to worry about consensus or client tables etc. but rather containing all that complexity in one place within VSR.

This not only makes it more friendly as a library, but reduces the probability of a correctness/liveness bug due to divergent logic between the StateMachine-under-test with the VOPR and what we actually run in production.